File: practicas/practica_1/Deepspace_Ruby_P1/lib/Weapon.rb
#encoding: utf-8

# File: Weapon.rb
# Author: Juan Manuel Rodríguez Gómez

module Deepspace

    # Brief: Representa a las armas de las que puede disponer una estación espacial 
    #        para potenciar su energía al disparar
    class Weapon

        # Constructor con parámetros
        def initialize(p_name, p_type, p_uses)
            @name = p_name
            @type = p_type
            @uses = p_uses
        end

        # Constructor de copia
        def self.newCopy(w)
            copy = Weapon.new(w.name, w.type, w.uses)
        end

        public

        #Consultor del tipo de arma
        def type
            if(@type == WeaponType::LASER)
                return WeaponType::LASER
            elsif (@type == WeaponType::MISSILE)
                return WeaponType::MISSILE
            else
                return WeaponType::PLASMA
            end
        end
      
        #Consultor del nombre
        def name
            @name
        end

        #Consultor de los usos
        def uses
            @uses
        end

        # Devuelve la potencia de disparo indicada por el tipo de arma.
        def power
            return type.power
        end

        # Emula el uso del arma.
        def useIt
            if (uses > 0)
              @uses -= 1
              return power
            else
              return 1.0
            end
        end
    end
end

File: practicas/practica_1/deepspace_Java_P1/src/main/java/deepspace/Weapon.java
/**
 * @file Weapon.java
 * @author Juan Manuel Rodríguez Gómez
 */
package deepspace;

/**
 * @brief Representa a las armas de las que puede disponer una estación espacial 
 *        para potenciar su energía al disparar
 */
class Weapon {

    private String name;
    private WeaponType type;
    private int uses;
    
    /**
     * @brief Constructor con parámetros
     * @param p_name Valor del nombre
     * @param p_type Valor del tipo de arma
     * @param p_uses Valor de los usos
     */
    Weapon(String p_name, WeaponType p_type, int p_uses) {
        name = p_name;
        type = p_type;
        uses = p_uses;
    }
    
    /**
     * @brief Constructor de copia
     * @param w Otro objeto de la clase Weapon
     */
    Weapon(Weapon w) {
        name = w.name;
        type = w.type;
        uses = w.uses;
    }
    
    /**
     * @brief Consultor del tipo de arma
     * @return type
     */
    public final WeaponType getType() {
        return type;
    }
    
    /**
     * @brief Consultor de los usos
     * @return uses
     */
    public final int getUses() {
        return uses;
    }

    /**
     * @brief Devuelve la potencia de disparo indicada por el tipo de arma.
     */
    public final float power() {
        return type.getPower();
    }
    
    /**
     * @brief Emula el uso del arma.
     */
    public final float useIt() {
        if(getUses() > 0) {
            uses--;
            return power();
        }
        else
            return 1.0f;
    }
}